crates/ruff_linter/src/message/mod.rs

```diff
@@ -34,46 +39,192 @@ mod rdjson;
 mod sarif;
 mod text;
 
+/// Message represents either a diagnostic message corresponding to a rule violation or a syntax
+/// error message raised by the parser.
 #[derive(Debug, PartialEq, Eq)]
-pub struct Message {
+pub enum Message {
+    Diagnostic(DiagnosticMessage),
+    SyntaxError(SyntaxErrorMessage),
+}
+
+/// A diagnostic message corresponding to a rule violation.
+#[derive(Debug, PartialEq, Eq)]
+pub struct DiagnosticMessage {
     pub kind: DiagnosticKind,
     pub range: TextRange,
     pub fix: Option<Fix>,
     pub file: SourceFile,
     pub noqa_offset: TextSize,
 }
 
+/// A syntax error message raised by the parser.
+#[derive(Debug, PartialEq, Eq)]
+pub struct SyntaxErrorMessage {
+    pub message: String,
+    pub range: TextRange,
+    pub file: SourceFile,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum MessageKind {
+    Diagnostic(Rule),
+    SyntaxError,
+}
+
+impl MessageKind {
+    pub fn as_str(&self) -> &str {
+        match self {
+            MessageKind::Diagnostic(rule) => rule.as_ref(),
+            MessageKind::SyntaxError => "syntax-error",
+        }
+    }
+}
+
 impl Message {
+    /// Create a [`Message`] from the given [`Diagnostic`] corresponding to a rule violation.
     pub fn from_diagnostic(
         diagnostic: Diagnostic,
         file: SourceFile,
         noqa_offset: TextSize,
-    ) -> Self {
-        Self {
+    ) -> Message {
+        Message::Diagnostic(DiagnosticMessage {
             range: diagnostic.range(),
             kind: diagnostic.kind,
             fix: diagnostic.fix,
             file,
             noqa_offset,
+        })
+    }
+
+    /// Create a [`Message`] from the given [`ParseError`].
+    pub fn from_parse_error(
+        parse_error: &ParseError,
+        locator: &Locator,
+        file: SourceFile,
+    ) -> Message {
+        // Try to create a non-empty range so that the diagnostic can print a caret at the right
+        // position. This requires that we retrieve the next character, if any, and take its length
+        // to maintain char-boundaries.
+        let len = locator
```

Comment:
Uh, our parse errors only have a start offset and not a range. That's funny.

Reply:
I'm not sure I follow here. This is mainly a copy-paste from the existing logic :)